src/proxy/http/balance.rs

```diff
@@ -118,19 +152,471 @@ where
 impl<F, A, B> Future for MakeSvc<F, A, B>
 where
     F: Future,
-    F::Item: Discover,
+    F::Item: Discover + HasEndpointStatus,
     <F::Item as Discover>::Service: svc::Service<http::Request<A>, Response = http::Response<B>>,
     A: Payload,
     B: Payload,
 {
-    type Item = Balance<WithPeakEwma<F::Item, PendingUntilFirstData>, PowerOfTwoChoices>;
+    type Item = Service<Balance<WithPeakEwma<F::Item, PendingUntilFirstData>, PowerOfTwoChoices>>;
     type Error = F::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         let discover = try_ready!(self.inner.poll());
+        let status = discover.endpoint_status();
         let instrument = PendingUntilFirstData::default();
         let loaded = WithPeakEwma::new(discover, self.default_rtt, self.decay, instrument);
-        Ok(Balance::p2c(loaded).into())
+        let balance = Balance::p2c(loaded);
+        Ok(Async::Ready(Service { balance, status }))
+    }
+}
+
+impl<S, A, B> svc::Service<http::Request<A>> for Service<S>
+where
+    S: svc::Service<http::Request<A>, Response = http::Response<B>>,
+    S::Error: Into<Error>,
+{
+    type Response = http::Response<B>;
+    type Error = Error;
+    type Future = future::Either<
+        future::FutureResult<Self::Response, Self::Error>,
+        future::MapErr<S::Future, fn(S::Error) -> Self::Error>,
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.balance.poll_ready().map_err(Into::into)
+    }
+
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
+        if self.status.is_empty() {
+            return future::Either::A(future::err(Box::new(NoEndpoints { _p: () })));
+        }
+        future::Either::B(self.balance.call(req).map_err(Into::into))
+    }
+}
+
+impl<S> HasEndpointStatus for Service<S> {
+    fn endpoint_status(&self) -> EndpointStatus {
+        self.status.clone()
+    }
+}
+
+impl fmt::Display for NoEndpoints {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt("load balancer has no endpoints", f)
+    }
+}
+
+impl ::std::error::Error for NoEndpoints {}
+
+pub mod fallback {
```

Comment:
this module doubles the size of this file... can it be split into its own file? maybe `balance` needs to become a directory...

Reply:
yeah, I wasn't sure --- will split it out!